### PR TITLE
feat(server): support wildcard business recognition emails

### LINKED ISSUE #3811: Enable wildcard on business suggestion_data emails
Some suppliers send invoices with unique address per invoice, eg `qr45uf@cloudflare.com`<br>Enable setting the recognition emails to accept and handle `*.cloudflare.com`

### PR DESCRIPTION
**Pull request by gilgardosh**
## What & why

Closes #3811.

Some suppliers send invoices from a **unique address per invoice** (e.g. `qr45uf@cloudflare.com`). Today a business is recognized only by an exact address in `suggestion_data.emails`, so these senders can never be matched.

This PR lets `suggestion_data.emails` store **wildcard patterns** (e.g. `*@cloudflare.com`) alongside concrete addresses, where `*` matches any run of characters. A single entry now recognizes all of a supplier's per-invoice addresses.

## Changes

- **Schema** (`business-suggestion-data-schema.helper.ts`): recognition emails now accept an email-shaped wildcard pattern (`*` allowed anywhere) in addition to a concrete address. Over-broad values such as a bare `*` are still rejected so a pattern can't accidentally match every sender.
- **New helper** `email-pattern.helper.ts`: the single source of truth for the pattern shape, plus an in-process, case-insensitive matcher used by tests and non-DB call sites.
- **Recognition lookups**: both `BusinessesProvider.getBusinessByEmail` and the email-ingestion control provider's ingest query translate `*` to a SQL `LIKE` wildcard, escaping the `LIKE` metacharacters `_`, `%` and `\` so a literal entry keeps matching exactly. Matching stays case-insensitive.
- **Tests**: unit tests for the matcher and pattern validation, schema tests for wildcard acceptance/rejection, and a control-provider test proving `qr45uf@cloudflare.com` is recognized via a stored `*@cloudflare.com` pattern.

## Matching semantics

`*` maps to `.*` (regex) / `%` (SQL `LIKE`); everything else is literal — `.`, `_`, `+` in an address are **not** treated as wildcards. The SQL translation was verified to be equivalent to the in-process matcher across literals, wildcards, and metacharacter edge cases.

## Notes

CI runs the full generate/lint/test pipeline. The test suite couldn't be run in the authoring environment because dependency install is blocked by an egress-policy denial for `cdn.sheetjs.com` (a transitive `node-xlsx` asset). The pure matching/validation logic and the SQL⇔JS equivalence were validated with standalone Node checks.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01MBmE2eA6XU1Nk5j1f9pmjt

---
_Generated by [Claude Code](https://claude.ai/code/session_01MBmE2eA6XU1Nk5j1f9pmjt)_

### CODE REVIEW

File: packages/server/src/modules/financial-entities/providers/businesses.provider.ts [hunk 1]
```diff
@@ -31,13 +31,31 @@ const getAllBusinesses = sql<IGetAllBusinessesQuery>`
...
+        FROM jsonb_array_elements_text(
+          CASE WHEN jsonb_typeof(b.suggestion_data->'emails') = 'array'
+               THEN b.suggestion_data->'emails'
+               ELSE '[]'::jsonb
+          END
+        ) AS candidate
+       WHERE lower($email::text) LIKE
+         replace(replace(replace(replace(lower(candidate), '\\', '\\\\'), '%', '\\%'), '_', '\\_'), '*', '%')
+         ESCAPE '\\'
+    )
```
[the remark this answers is not in the document]
Comment (gilgardosh) [reply packages/server/src/modules/financial-entities/providers/businesses.provider.ts:62]: Not applying this one — the premise doesn't hold here. There is no GIN index on `accounter_schema.businesses.suggestion_data` anywhere in the migrations (the only GIN indexes in the schema are trigram indexes on `charges`/`transactions`/`documents`). So the original `? $email` query was already a sequential scan; this change doesn't regress from an indexed lookup to a scan.

Given that, the suggested `(? ...) OR EXISTS (...)` fast-path wouldn't help and would slightly hurt: `OR` short-circuits only for the single *matching* row, so every non-matching row (the vast majority) evaluates `?` (false) **and then** the `EXISTS` unnest anyway — strictly more per-row work than the current single `EXISTS`.

If exact-match throughput on this table ever becomes a concern, the right fix is to add a `GIN (suggestion_data)` index and then reintroduce the hybrid `?`-first predicate — but the `businesses` table is small (per-tenant, scoped by RLS) and this lookup only runs on inbound email ingestion, so I don't think it's warranted now.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/server/src/modules/financial-entities/helpers/email-pattern.helper.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+const WILDCARD_EMAIL_SHAPE = /^[^@\s]+@[^@\s]+\.[^@\s]+$/;
+
+/**
+ * Validate a wildcard recognition entry. Non-wildcard entries are plain email
+ * addresses and are validated elsewhere (the suggestion-data schema); this only
+ * covers entries that actually contain a `*`.
+ */
+export function isValidWildcardEmailPattern(value: string): boolean {
+  return isWildcardEmailPattern(value) && WILDCARD_EMAIL_SHAPE.test(value.trim());
+}
```
[the remark this answers is not in the document]
Comment (gilgardosh) [reply packages/server/src/modules/financial-entities/helpers/email-pattern.helper.ts:47]: Good catch — applied in b7e0f22. `isValidWildcardEmailPattern` now requires the domain to carry at least one concrete label (2+ chars, no `*`) before the TLD, so `*@*.com` and `*@*.*` are rejected while `*@cloudflare.com` and `*@*.cloudflare.com` stay valid. Added tests for the tightened rule in both `email-pattern.helper.test.ts` and the schema test.

---
_Generated by [Claude Code](https://claude.ai/code)_
